Distances returned by calling `search` is inaccurate

**Please state you issue using the following template and, most importantly, in English.**

**Describe the bug**
```
self = <test_search.TestSearchBase object at 0x7fc91429f940>, connect = <milvus.client.stub.Milvus object at 0x7fc91746d470>, collection = 'search_B9A9zjBK', get_simple_index = {'index_type': 'ANNOY', 'n_trees': 4}

    def test_search_distance_l2_after_index(self, connect, collection, get_simple_index):
        '''
        target: search collection, and check the result: distance
        method: compare the return distance value with value computed with Inner product
        expected: the return distance equals to the computed value
        '''
        index_type = get_simple_index["index_type"]
        nq = 2
        entities, ids = init_data(connect, collection)
        connect.create_index(collection, field_name, default_index_name, get_simple_index)
        search_param = get_search_param(index_type)
        query, vecs = gen_query_vectors_rand_entities(field_name, entities, top_k, nq, search_params=search_param)
        inside_vecs = entities[-1]["values"]
        min_distance = 1.0
        for i in range(nb):
            tmp_dis = l2(vecs[0], inside_vecs[i])
            if min_distance > tmp_dis:
                min_distance = tmp_dis
        res = connect.search(collection, query)
>       assert abs(np.sqrt(res[0]._distances[0]) - min_distance) <= gen_inaccuracy(res[0]._distances[0])
E       AssertionError: assert 0.19996865597885105 <= 0.0024112313401465323
E        +  where 0.19996865597885105 = abs((0.784132636572006 - 0.584163980593155))
E        +    where 0.784132636572006 = <ufunc 'sqrt'>(0.6148639917373657)
E        +      where <ufunc 'sqrt'> = np.sqrt
E        +  and   0.0024112313401465323 = gen_inaccuracy(0.6148639917373657)

test_search.py:504: AssertionError


```

both in l2/ip

**Steps/Code to reproduce behavior**
Follow this [guide](http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) to craft a minimal bug report. This helps us reproduce the issue you're having and resolve the issue more quickly.

**Expected behavior**
A clear and concise description of what you expected to happen.

**Environment details**
- Hardware/Softward conditions (OS, CPU, GPU, Memory)
- Method of installation (Docker, or from source)
- Milvus version (v0.3.1, or v0.4.0)
- Milvus configuration (Settings you made in `server_config.yaml`)

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -28,6 +28,7 @@ Please mark all changes in change log and use the issue from GitHub
 -   \#2763 Unexpected error when insert binary entities
 -   \#2765 Server crashed when calling `get_entity_by_id`
 -   \#2783 Wrong result returned if searching with tags
+-   \#2790 Distances returned by calling `search` is inaccurate
 
 ## Feature
 -   \#2319 Redo metadata to support MVCC

--- a/core/src/db/DBImpl.cpp
+++ b/core/src/db/DBImpl.cpp
@@ -1773,6 +1773,7 @@ DBImpl::CreateIndex(const std::shared_ptr<server::Context>& context, const std::
 
         // step 2: check index difference
         CollectionIndex old_index;
+        old_index.field_name_ = index.field_name_;
         status = DescribeIndex(collection_id, old_index);
         if (!status.ok()) {
             LOG_ENGINE_ERROR_ << "Failed to get collection index info for collection: " << collection_id;
@@ -1781,6 +1782,11 @@ DBImpl::CreateIndex(const std::shared_ptr<server::Context>& context, const std::
 
         // step 3: update index info
         CollectionIndex new_index = index;
+        json new_index_json = new_index.extra_params_;
+        if (new_index_json.contains("index_type")) {
+            new_index.engine_type_ = (int32_t)engine::s_map_engine_type.at(new_index_json["index_type"]);
+        }
+
         new_index.metric_type_ = old_index.metric_type_;  // dont change metric type, it was defined by CreateCollection
         if (!utils::IsSameIndex(old_index, new_index)) {
             status = UpdateCollectionIndexRecursively(collection_id, new_index);

--- a/core/src/db/Types.h
+++ b/core/src/db/Types.h
@@ -37,6 +37,8 @@ typedef std::vector<faiss::Index::idx_t> ResultIds;
 typedef std::vector<faiss::Index::distance_t> ResultDistances;
 
 struct CollectionIndex {
+    std::string field_name_;
+    std::string index_name_;
     int32_t engine_type_ = (int)EngineType::FAISS_IDMAP;
     int32_t metric_type_ = (int)MetricType::L2;
     milvus::json extra_params_ = {{"nlist", 2048}};

--- a/core/src/db/meta/SqliteMetaImpl.cpp
+++ b/core/src/db/meta/SqliteMetaImpl.cpp
@@ -848,6 +848,17 @@ SqliteMetaImpl::DescribeCollectionIndex(const std::string& collection_id, Collec
         } else {
             return Status(DB_NOT_FOUND, "Collection " + collection_id + " not found");
         }
+
+        //        auto field_groups =
+        //            ConnectorPtr->select(columns(&hybrid::FieldSchema::index_name_,
+        //            &hybrid::FieldSchema::index_param_),
+        //                                 where(c(&hybrid::FieldSchema::collection_id_) == collection_id and
+        //                                       c(&hybrid::FieldSchema::field_name_) == index.field_name_));
+        //
+        //        if (groups.size() == 1) {
+        //            index.index_name_ = std::get<0>(groups[0]);
+        //            index.extra_params_ = milvus::json::parse(std::get<1>(groups[1]));
+        //        }
     } catch (std::exception& e) {
         return HandleException("Encounter exception when describe index", e.what());
     }

--- a/core/src/server/ValidationUtil.cpp
+++ b/core/src/server/ValidationUtil.cpp
@@ -25,12 +25,11 @@ namespace server {
 
 namespace {
 
-constexpr size_t COLLECTION_NAME_SIZE_LIMIT = 255;
+constexpr size_t NAME_SIZE_LIMIT = 255;
 constexpr int64_t COLLECTION_DIMENSION_LIMIT = 32768;
 constexpr int32_t INDEX_FILE_SIZE_LIMIT = 4096;  // index trigger size max = 4096 MB
 constexpr int64_t M_BYTE = 1024 * 1024;
 constexpr int64_t MAX_INSERT_DATA_SIZE = 256 * M_BYTE;
-constexpr int64_t FIELD_NAME_SIZE_LIMIT = 255;
 
 Status
 CheckParameterRange(const milvus::json& json_params, const std::string& param_name, int64_t min, int64_t max,
@@ -102,7 +101,7 @@ ValidateCollectionName(const std::string& collection_name) {
 
     std::string invalid_msg = "Invalid collection name: " + collection_name + ". ";
     // Collection name size shouldn't exceed 255.
-    if (collection_name.size() > COLLECTION_NAME_SIZE_LIMIT) {
+    if (collection_name.size() > NAME_SIZE_LIMIT) {
         std::string msg = invalid_msg + "The length of a collection name must be less than 255 characters.";
         LOG_SERVER_ERROR_ << msg;
         return Status(SERVER_INVALID_COLLECTION_NAME, msg);
@@ -140,7 +139,7 @@ ValidateFieldName(const std::string& field_name) {
 
     std::string invalid_msg = "Invalid field name: " + field_name + ". ";
     // Field name size shouldn't exceed 255.
-    if (field_name.size() > FIELD_NAME_SIZE_LIMIT) {
+    if (field_name.size() > NAME_SIZE_LIMIT) {
         std::string msg = invalid_msg + "The length of a field name must be less than 255 characters.";
         LOG_SERVER_ERROR_ << msg;
         return Status(SERVER_INVALID_FIELD_NAME, msg);
@@ -157,6 +156,44 @@ ValidateFieldName(const std::string& field_name) {
     int64_t field_name_size = field_name.size();
     for (int64_t i = 1; i < field_name_size; ++i) {
         char name_char = field_name[i];
+        if (name_char != '_' && std::isalnum(name_char) == 0) {
+            std::string msg = invalid_msg + "Field name cannot only contain numbers, letters, and underscores.";
+            LOG_SERVER_ERROR_ << msg;
+            return Status(SERVER_INVALID_FIELD_NAME, msg);
+        }
+    }
+
+    return Status::OK();
+}
+
+Status
+ValidateIndexName(const std::string& index_name) {
+    // Index name shouldn't be empty.
+    if (index_name.empty()) {
+        std::string msg = "Index name should not be empty.";
+        LOG_SERVER_ERROR_ << msg;
+        return Status(SERVER_INVALID_FIELD_NAME, msg);
+    }
+
+    std::string invalid_msg = "Invalid index name: " + index_name + ". ";
+    // Index name size shouldn't exceed 255.
+    if (index_name.size() > NAME_SIZE_LIMIT) {
+        std::string msg = invalid_msg + "The length of a field name must be less than 255 characters.";
+        LOG_SERVER_ERROR_ << msg;
+        return Status(SERVER_INVALID_FIELD_NAME, msg);
+    }
+
+    // Field name first character should be underscore or character.
+    char first_char = index_name[0];
+    if (first_char != '_' && std::isalpha(first_char) == 0) {
+        std::string msg = invalid_msg + "The first character of a field name must be an underscore or letter.";
+        LOG_SERVER_ERROR_ << msg;
+        return Status(SERVER_INVALID_FIELD_NAME, msg);
+    }
+
+    int64_t field_name_size = index_name.size();
+    for (int64_t i = 1; i < field_name_size; ++i) {
+        char name_char = index_name[i];
         if (name_char != '_' && std::isalnum(name_char) == 0) {
             std::string msg = invalid_msg + "Field name cannot only contain numbers, letters, and underscores.";
             LOG_SERVER_ERROR_ << msg;
@@ -458,7 +495,7 @@ ValidatePartitionName(const std::string& partition_name) {
 
     std::string invalid_msg = "Invalid partition name: " + partition_name + ". ";
     // Collection name size shouldn't exceed 255.
-    if (partition_name.size() > COLLECTION_NAME_SIZE_LIMIT) {
+    if (partition_name.size() > NAME_SIZE_LIMIT) {
         std::string msg = invalid_msg + "The length of a partition name must be less than 255 characters.";
         LOG_SERVER_ERROR_ << msg;
         return Status(SERVER_INVALID_COLLECTION_NAME, msg);

--- a/core/src/server/ValidationUtil.h
+++ b/core/src/server/ValidationUtil.h
@@ -31,6 +31,9 @@ extern Status
 ValidateFieldName(const std::string& field_name);
 
 extern Status
+ValidateIndexName(const std::string& index_name);
+
+extern Status
 ValidateTableDimension(int64_t dimension, int64_t metric_type);
 
 extern Status

--- a/core/src/server/delivery/request/CreateIndexRequest.cpp
+++ b/core/src/server/delivery/request/CreateIndexRequest.cpp
@@ -54,10 +54,21 @@ CreateIndexRequest::OnExecute() {
             return status;
         }
 
+        status = ValidateFieldName(field_name_);
+        if (!status.ok()) {
+            return status;
+        }
+
+        status = ValidateIndexName(index_name_);
+        if (!status.ok()) {
+            return status;
+        }
+
         // only process root collection, ignore partition collection
         engine::meta::CollectionSchema collection_schema;
+        engine::meta::hybrid::FieldsSchema fields_schema;
         collection_schema.collection_id_ = collection_name_;
-        status = DBWrapper::DB()->DescribeCollection(collection_schema);
+        status = DBWrapper::DB()->DescribeHybridCollection(collection_schema, fields_schema);
         fiu_do_on("CreateIndexRequest.OnExecute.not_has_collection",
                   status = Status(milvus::SERVER_UNEXPECTED_ERROR, ""));
         fiu_do_on("CreateIndexRequest.OnExecute.throw_std.exception", throw std::exception());
@@ -109,6 +120,10 @@ CreateIndexRequest::OnExecute() {
 
         // step 3: create index
         engine::CollectionIndex index;
+        if (json_params_.contains("metric_type")) {
+            index.metric_type_ = (int32_t)engine::s_map_metric_type.at(json_params_["metric_type"]);
+        }
+
         index.engine_type_ = adapter_index_type;
         index.extra_params_ = json_params_;
         status = DBWrapper::DB()->CreateIndex(context_, collection_name_, index);

